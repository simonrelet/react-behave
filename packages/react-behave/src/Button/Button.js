import PropTypes from 'prop-types';
import React from 'react';

function Button({ children, ...otherProps }) {
  return <button {...otherProps}>{children}</button>;
}

Button.propTypes = {
  children: PropTypes.node.isRequired,
};

export default Button;
