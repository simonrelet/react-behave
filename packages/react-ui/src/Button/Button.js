import { toInnerRef } from '@simonrelet/react-utils';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
// Side effect import as the Button component apply CSS classes on the SVGIcon
import '../icons/SVGIcon';
import injectStyles from '../injectStyles';

function styles({ borderRadius, palette, relativeSpacing }) {
  return {
    root: {
      alignItems: 'center',
      backgroundColor: 'transparent',
      border: {
        width: 1,
        style: 'solid',
        color: palette.grey,
      },
      borderRadius,
      cursor: 'pointer',
      display: 'inline-flex',
      fontFamily: 'inherit',
      fontSize: '1rem',
      outline: 'none',
      padding: ({ renderIconLeft, renderIconRight }) => {
        return [
          relativeSpacing(),
          relativeSpacing(renderIconRight ? 1 : 2),
          relativeSpacing(),
          relativeSpacing(renderIconLeft ? 1 : 2),
        ].join(' ');
      },
      transition: 'all 120ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',

      '&:focus': {
        borderColor: palette.primary,
        color: palette.primary,
      },

      '&:hover:not(:focus)': {
        borderColor: palette.emphasize(palette.grey),
      },
    },
    dense: {
      fontSize: '0.875rem',
    },
    content: {
      flex: 'none',
    },
    icon: {
      flex: 'none',
      transition: 'transform 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
    },
    iconLeft: {
      composes: '$icon',
      marginRight: relativeSpacing(),
    },
    iconRight: {
      composes: '$icon',
      marginLeft: relativeSpacing(),
    },
  };
}

function Button({
  children,
  classes,
  dense,
  innerRef,
  renderIconLeft,
  renderIconRight,
  ...otherProps
}) {
  return (
    <button
      {...otherProps}
      className={classNames(classes.root, { [classes.dense]: dense })}
      ref={innerRef}
    >
      {renderIconLeft && renderIconLeft({ className: classes.iconLeft })}
      <span className={classes.content}>{children}</span>
      {renderIconRight && renderIconRight({ className: classes.iconRight })}
    </button>
  );
}

Button.propTypes = {
  children: PropTypes.node.isRequired,
  classes: PropTypes.object.isRequired,
  dense: PropTypes.bool,
  innerRef: PropTypes.any,
  renderIconLeft: PropTypes.func,
  renderIconRight: PropTypes.func,
};

Button.defaultProps = {
  dense: false,
};

export default injectStyles(styles)(toInnerRef(Button));
