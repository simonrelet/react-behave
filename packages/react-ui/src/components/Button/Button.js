import {
  darken,
  getLuminance,
  lighten,
  readableColor,
  transparentize,
} from 'polished'
import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'

function emphasize(color) {
  return getLuminance(color) > 0.5 ? darken(0.15, color) : lighten(0.15, color)
}

const defaultColor = 'white'

const attrs = {
  color: p => p.theme.palette[p.color || defaultColor],

  padding: p => [
    p.theme.relativeSpacing(),
    p.theme.relativeSpacing(p.withIconRight ? 1 : 2),
    p.theme.relativeSpacing(),
    p.theme.relativeSpacing(p.withIconLeft ? 1 : 2),
  ],
}

const StyledButton = styled.button.attrs(attrs)`
  align-items: center;
  background-color: ${p => p.color};
  border: none;
  border-radius: ${p => p.theme.borderRadius};
  color: ${p => readableColor(p.color)};
  cursor: pointer;
  display: inline-flex;
  font-family: inherit;
  font-size: ${p => (p.dense ? '0.875rem' : '1rem')};
  outline: none;
  padding: ${p => p.padding.join(' ')};
  transition: all 120ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;

  &:focus {
    box-shadow: 0 0 0 0.2rem
      ${p =>
        transparentize(
          0.5,
          p.color === p.theme.palette.white ? p.theme.palette.primary : p.color,
        )};
  }

  &:active {
    background-color: ${p => emphasize(p.color)};
    color: ${p => readableColor(emphasize(p.color))};
  }
`

const StyledContent = styled.span`
  margin-left: ${p => (p.withIconLeft ? p.theme.relativeSpacing() : '0')}
  margin-right: ${p => (p.withIconRight ? p.theme.relativeSpacing() : '0')}
`

function Button({
  children,
  color,
  dense,
  innerRef,
  renderIconLeft,
  renderIconRight,
  ...otherProps
}) {
  const iconLeft = renderIconLeft()
  const iconRight = renderIconRight()

  return (
    <StyledButton
      {...otherProps}
      innerRef={innerRef}
      color={color}
      dense={dense}
      withIconLeft={!!iconLeft}
      withIconRight={!!iconRight}
    >
      {iconLeft}
      <StyledContent withIconLeft={!!iconLeft} withIconRight={!!iconRight}>
        {children}
      </StyledContent>
      {iconRight}
    </StyledButton>
  )
}

Button.propTypes = {
  children: PropTypes.node.isRequired,
  color: PropTypes.string,
  dense: PropTypes.bool,
  innerRef: PropTypes.any,
  renderIconLeft: PropTypes.func,
  renderIconRight: PropTypes.func,
}

Button.defaultProps = {
  color: defaultColor,
  dense: false,
  renderIconLeft: () => null,
  renderIconRight: () => null,
}

export default Button
