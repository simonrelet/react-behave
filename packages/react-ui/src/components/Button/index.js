export { default, severities } from './Button';

