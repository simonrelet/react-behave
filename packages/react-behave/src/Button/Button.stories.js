import { action } from '@storybook/addon-actions';
import { boolean, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import Button from './Button';

const stories = storiesOf('Button', module);

stories.add('Basic button', () => (
  <Button disabled={boolean('Disabled', false)} onClick={action('onClick')}>
    {text('Label', 'Click me')}
  </Button>
));
