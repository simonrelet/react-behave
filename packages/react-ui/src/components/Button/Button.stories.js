import { action } from '@storybook/addon-actions';
import { boolean, select, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import changeCase from 'change-case';
import React from 'react';
import styled from 'styled-components';
import { allColors, color } from '../../../storybook/knobs';
import { allIcons } from '../../icons';
import Button from './Button';

const stories = storiesOf('Button', module);

const iconsOptions = ['None', ...allIcons.map(icon => icon.name)];

function renderIcon(side) {
  return () => {
    const iconName = select(`${side} icon`, iconsOptions, 'None');
    const Icon = allIcons.find(icon => icon.name === iconName);
    return Icon ? <Icon.component /> : null;
  };
}

const Row = styled.div`
  display: flex;
  align-items: center;
  & > *:not(:first-child) {
    margin-left: ${p => p.theme.spacing()};
  }
`;

stories.add('Button colors', () => (
  <Row>
    {allColors.map(c => (
      <Button key={c} onClick={action('onClick')} color={c}>
        {changeCase.sentenceCase(c)}
      </Button>
    ))}
  </Row>
));

stories.add('Dynamic button', () => (
  <Button
    dense={boolean('Dense', false)}
    onClick={action('onClick')}
    renderIconLeft={renderIcon('Left')}
    renderIconRight={renderIcon('Right')}
    color={color('Color')}
  >
    {text('Label', 'Click me')}
  </Button>
));
