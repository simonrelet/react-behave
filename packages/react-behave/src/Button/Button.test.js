import React from 'react';
import renderer from 'react-test-renderer';
import Button from './Button';

describe('<Button />', () => {
  it('renders correctly', () => {
    const props = { a: 0, b: 'b' };
    const tree = renderer.create(<Button {...props}>Click me</Button>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
