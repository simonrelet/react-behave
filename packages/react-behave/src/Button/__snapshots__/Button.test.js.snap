// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders correctly 1`] = `
<button
  a={0}
  b="b"
>
  Click me
</button>
`;
